Add missing task events

joshuatf opened the pull request:
* Adds missing events
* Adds traits to handle prefixing and recording events in task/task lists

@pmcpinto If you want to give this list a glance to make sure we've covered everything that would be great!  No changes were made to events inside individual tasks, so these events just cover general task/task list events.

### Detailed test instructions:

1. Add an error log to `wc_admin_record_tracks_event()` in `includes/core-functions.php`:
```php
error_log( $event_name );
```
2. Enable the logging of Tracks events to your browser dev console `localStorage.setItem( 'debug', 'wc-admin:tracks' );`
3. Under one of the tasks in `src/Features/OnboardingTasks/Tasks/*.php` add `'is_dismissable' => true;` and `'is_snoozeable' => true`
3. Check to make sure the following events can all be triggered with the correct task names:

(C) - denotes client-side events, found in your console
(S) - denotes server-side events, found in debug log

`tasklist_task_completed` - (S) Complete any task in the list, on or off the homepage.  Note you may have to revisit the task list if completing on a different page.
`tasklist_click` - (C) View the task list and click on any task in the list
`tasklist_dismiss_task` - (S) Dismiss a task via the dropdown
`tasklist_undo_remindmelater_task` - (S) Undo task dismissal via the transient notice
`tasklist_remindmelater_task` - (S) Snooze any task via dropdown
`tasklist_undo_dismiss_task` - (S) Undo the snoozing of a task via transient notice
`task_view` - (C) View a single task on the frontend
`tasklist_view` - (C) View any of the task lists on the frontend
`tasklist_completed` - (S) Hide the task list via the dropdown
`tasklist_toggled` - (S) Should be triggered when toggling the tasklist on/off under "Screen Options" on core Orders screen.  I'm missing the toggles in my screen options (also in previous versions) - can anyone replicate?
`tasklist_skip` - (S) Seems to have been removed, but could use a sanity check.

These tasks are part of https://github.com/woocommerce/woocommerce-admin/pull/7736 :
`tasklist_tasks_completed` - (S) Complete all visible tasks in the task list
`extended_tasklist_tasks_completed` - (S) Complete all visible tasks in the extended task list

joshuatf committed:
Add task view event

joshuatf committed:
Fix double prefix on events

joshuatf committed:
Add shared trait for event recording

joshuatf commented on packages/onboarding/src/components/WooOnboardingTask/WooOnboardingTask.js:
```diff
@@ -1,16 +1,48 @@
-	<Fill name={ 'woocommerce_onboarding_task_' + id } { ...props } />
-);
+export const trackView = ( taskId ) => {
+	const activePlugins = wp.data
```
Note that these selectors are normally async, but currently we preload this data.  I think a better solution may be warranted long-term, but this is an existing issue and currently does not create any bugs.

joshuatf committed:
Fix double prefixed events

joshuatf committed:
Use record_tracks_event to record task related events

joelclimbsthings commented:
Hey @joshuatf , well done! I went over all of the listed events, and here are my notes:

Some of the events include the task ID within the name, and others didn't. I'm a little uncertain of when this should be the case so I thought I'd mention it. I'll list them in more detail below.

Client-side events are prepended with `wcadmin_` and server-side are not. I'm assuming this is addressed further down in the stack on the server-side, but thought I'd mention it as well. 

Most notably I had an error thrown when trying to fire `tasklist_tasks_completed` after all tasks on the setup list were completed:
```
Fatal error:  Uncaught ArgumentCountError: Too few arguments to function Automattic\WooCommerce\Admin\Features\OnboardingTasks\TaskList::record_tracks_event(), 1 passed in /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskList.php on line 190 and exactly 2 expected in /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskTraits.php:34
Stack trace:
#0 /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskList.php(190): Automattic\WooCommerce\Admin\Features\OnboardingTasks\TaskList->record_tracks_event('tasks_completed')
#1 /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskList.php(199): Automattic\WooCommerce\Admin\Features\OnboardingTasks\TaskList->possibly_track_completion()
#2 /home/joel in /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskTraits.php on line 34
```

Here is an event-by-event detail of what I saw:

<table>
<tr>
	<td>Expected name</td>
	<td>Observed name</td>
	<td>Notes</td>
</tr>
<tr>
	<td>tasklist_task_completed</td>
	<td>tasklist_store_details_task_completed</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_click</td>
	<td>wcadmin_tasklist_click</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_dismiss_task</td>
	<td>tasklist_appearance_dismiss_task</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_undo_dismiss_task</td>
	<td>tasklist_appearance_undo_dismiss_task</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_remindmelater_task</td>
	<td>tasklist_appearance_remindmelater_task</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_undo_remindmelater_task</td>
	<td>tasklist_appearance_undo_remindmelater_task</td>
	<td></td>
</tr>
<tr>
	<td>task_view</td>
	<td>wcadmin_task_view</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_view</td>
	<td>wcadmin_tasklist_view</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_completed</td>
	<td>tasklist_completed</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_toggled</td>
	<td></td>
	<td>I'm not able to see this under "screen options" either; I assume this was already broken?</td>
</tr>
<tr>
	<td>tasklist_skip</td>
	<td></td>
	<td>Didn't see any sign of this event, nor can I find it within the repo. Seems it's removed, as you mentioned?</td>
</tr>
<tr>
	<td>tasklist_tasks_completed</td>
	<td></td>
	<td>This resulted in the error mentioned above</td>
</tr>
<tr>
	<td>extended_tasklist_tasks_completed</td>
	<td></td>
	<td>Tried to fire this by running `npm run example -- --ext=add-task` as detailed in [this PR](https://github.com/woocommerce/woocommerce-admin/pull/7730), but wasn't able to complete the list. Is this functionality still in progress, and if not, how would I test it?</td>
</tr>
</table>

joelclimbsthings reviewed (changes requested):
Hey @joshuatf , well done! I went over all of the listed events, and here are my notes:

The server-side events include the task ID within the name, and client-side ones did not. I'm a little uncertain of whether this should be the case so I thought I'd mention it. I'll list them in more detail below.

Client-side events are also prepended with `wcadmin_` and server-side are not. I'm assuming this is addressed further down in the stack on the server-side, but thought I'd mention it as well. 

Most notably I had an error thrown when trying to fire `tasklist_tasks_completed` after all tasks on the setup list were completed:
```
Fatal error:  Uncaught ArgumentCountError: Too few arguments to function Automattic\WooCommerce\Admin\Features\OnboardingTasks\TaskList::record_tracks_event(), 1 passed in /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskList.php on line 190 and exactly 2 expected in /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskTraits.php:34
Stack trace:
#0 /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskList.php(190): Automattic\WooCommerce\Admin\Features\OnboardingTasks\TaskList->record_tracks_event('tasks_completed')
#1 /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskList.php(199): Automattic\WooCommerce\Admin\Features\OnboardingTasks\TaskList->possibly_track_completion()
#2 /home/joel in /home/joelt/Local Sites/mothrad/app/public/wp-content/plugins/woocommerce-admin-add-missing-task-events/src/Features/OnboardingTasks/TaskTraits.php on line 34
```

Here is an event-by-event detail of what I saw:

<table>
<tr>
	<td>Expected name</td>
	<td>Observed name</td>
	<td>Notes</td>
</tr>
<tr>
	<td>tasklist_task_completed</td>
	<td>tasklist_store_details_task_completed</td>
	<td>*Includes task ID</td>
</tr>
<tr>
	<td>tasklist_click (c)</td>
	<td>wcadmin_tasklist_click</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_dismiss_task</td>
	<td>tasklist_appearance_dismiss_task</td>
	<td>*Includes task ID</td>
</tr>
<tr>
	<td>tasklist_undo_dismiss_task</td>
	<td>tasklist_appearance_undo_dismiss_task</td>
	<td>*Includes task ID</td>
</tr>
<tr>
	<td>tasklist_remindmelater_task</td>
	<td>tasklist_appearance_remindmelater_task</td>
	<td>*Includes task ID</td>
</tr>
<tr>
	<td>tasklist_undo_remindmelater_task</td>
	<td>tasklist_appearance_undo_remindmelater_task</td>
	<td>*Includes task ID</td>
</tr>
<tr>
	<td>task_view (c)</td>
	<td>wcadmin_task_view</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_view (c)</td>
	<td>wcadmin_tasklist_view</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_completed</td>
	<td>tasklist_completed</td>
	<td></td>
</tr>
<tr>
	<td>tasklist_toggled</td>
	<td></td>
	<td>I'm not able to see this under "screen options" either; I assume this was already broken?</td>
</tr>
<tr>
	<td>tasklist_skip</td>
	<td></td>
	<td>Didn't see any sign of this event, nor can I find it within the repo. Seems it's removed, as you mentioned?</td>
</tr>
<tr>
	<td>tasklist_tasks_completed</td>
	<td></td>
	<td>This resulted in the error mentioned above</td>
</tr>
<tr>
	<td>extended_tasklist_tasks_completed</td>
	<td></td>
	<td>Tried to fire this by running `npm run example -- --ext=add-task` as detailed in [this PR](https://github.com/woocommerce/woocommerce-admin/pull/7730), but wasn't able to complete the list. Is this functionality still in progress, and if not, how would I test it?</td>
</tr>
</table>

joshuatf committed:
Handle PR feedback

joshuatf committed:
Add parent ID to temporary extended tasks

joshuatf committed:
Fix prefixing order

joshuatf commented:
Thanks for the feedback, @joelclimbsthings!  And sorry I was too hasty in pushing this after refactoring the recording; should have tested this more thoroughly.'

Fixed up events in https://github.com/woocommerce/woocommerce-admin/pull/7744/commits/892b47b11548a403be44f61a6e28cb234fe16f66.

>*Includes task ID

Added `parent_id` to tasks.  This now allows tasks to access their parent ID and record the correct events through traits.

>I'm not able to see this under "screen options" either; I assume this was already broken?

I finally found this again. It's under the "Help" tab on "Products" or "Orders" (core pages).  But you must have at least one product or order for them sh
![Screen Shot 2021-10-06 at 8 22 17 AM](https://user-images.githubusercontent.com/10561050/136202030-09b3bb75-fbde-453f-9eba-d2a8d3be24c4.png)
ow.

>This resulted in the error mentioned above

Fixed.

> extended_tasklist_tasks_completed. Is this functionality still in progress, and if not, how would I test it?

@louwie17 might be a better resource on testing this, but for easy testing, you can change one of the existing tasks to that task list.  `self::add_task( 'extended', Appearance::get_task() );`

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -785,6 +785,7 @@ public function dismiss_task( $request ) {
 				array(
 					'id'             => $id,
 					'is_dismissable' => true,
+					'parent_id'      => 'extended',
```
cc @louwie17 - Added the `parent_id` manually here since this is a temporary fix.  Just want to make sure that all dynamic tasks should belong to the `extended` task list- is that right?

joelclimbsthings reviewed (approved):
Thanks @joshuatf , changes look great! I was able to confirm most of the remaining events, including `tasklist_toggle`, `tasklist_tasks_completed` and `extended_tasklist_tasks_completed`. 

The only one remaining was `tasklist_skip`, but I've no idea what/when that would be fired, and I'm guessing it's obsolete. On that assumption consider this approved :shipit:
